accepted/future-releases/3009-inference-using-bounds/feature-specification.md

```diff
@@ -0,0 +1,319 @@
+  space:2
+  csstvUpwards["CSSTV"]
+  space
+  overallOutput["Output"]
+
+  space:8
+
+  space:4
+  constraintGeneration["X <: A<_>"]
+  space:3
+
+  space:2
+  boundsDownwards["X extends A<X>"]
+  space:2
+  boundsUpwards["X extends A <X>"]
+  space:2
+
+  downwards -- "∅" --> csstvDownwards
+  upwards -- "C <: X" --> csstvUpwards
+
+  overallInput -- "{X = _}" --> csstvDownwards
+  csstvDownwards -- "{X = _}" --> csstvUpwards
+  csstvUpwards -- "{X = C}" --> overallOutput
+
+  constraintGeneration --> newConstraint
+  boundsUpwards --> constraintGeneration
+
+  boundsDownwards --> csstvDownwards
+  boundsUpwards --> csstvUpwards
+```
+
+## Proposed changes
+
+We propose generating more constraints from existing constraints and the bound. The original example is solved by this proposal and a few more beneficial examples are discussed afterwards.
+
+In the motivating example, adding the best-effort bound approximation didn't affect the result since the set of constraints for `X` already contained a lower-bound constraint for `X` (`C <# X`), and that constraint had the priority over the upper-bound "best-effort" approximation constraint, since type `C` is fully known (i.e. doesn't contain `_`). In the current proposal ***we aim to derive a lower-bound constraint by combining the already existing lower-bound constraint with the actual bound***.
+
+In general, if we have two type constraints of the form `E <: Y` and `Y <: F`, where `Y` is a type variable and `E` and `F` are type schemas, it follows that `E <: F` holds if the `E <: Y` and `Y <: F` hold. So, what if instead of adding the constraint `X <: B'` in step 5 of CSSTV, where `B'` is the best-effort bound approximation, we generate a new set of constraints, running the subtype constraint generation algorithm described in [Subtype constraint generation][subtype-constraint-generation] for `P = C`, `Q = B`, `L = {..., X, ...}`.
+
+`C` doesn't contain type variables from `L` by construction, and `B` may contain type variables from `L` (confidently, `X` in case X is F-bounded, but also any type variables from the same declaration, regardless of F-boundness of `X`), which makes `C <# B` a valid input for [Subtype constraint generation][subtype-constraint-generation] and a good source of additional type information.
+
+The new constraints, generated by [Subtype constraint generation][subtype-constraint-generation] from `C <# B` are then added to the overall set of constraints for the current phase of type inference. Then, the solution for type variable `X` is computed from the new constraint set, at the end of the updated step 5 of CSSTV. 
+Note that when `X` isn't F-bounded and its bound doesn't refer to other type variables from the same declaration, the suggested changes for step 5 of CSSTV work exactly as the unmodified step 5 of the same algorithm.
+
+## Re-evaluating the motivating example
+Let's see how type inference, updated as proposed, works in the motivating example.
```

Comment:
I think an empty line would be expected here:
```suggestion

Let's see how type inference, updated as proposed, works in the motivating example.
```

Reply:
Thanks for the suggestion! I've updated the file accordingly.